### PR TITLE
bpo-36889: Document Stream class and add docstrings

### PR DESCRIPTION
**Pull request by tirkarthi**
* This just copies the docs from `StreamWriter` and `StreamReader`.
* Add docstring for asyncio functions.

https://bugs.python.org/issue36889

Automerge-Triggered-By: @asvetlov

### CODE REVIEW

File: Doc/library/asyncio-stream.rst [hunk 1]
```diff
@@ -367,13 +366,177 @@ Stream
...
+   :meth:`readline` returns empty data.
 
    Do not instantiate *Stream* objects directly; use API like :func:`connect`
    and :class:`StreamServer` instead.
 
    .. versionadded:: 3.8
 
+   .. attribute:: mode
+
+      Returns the mode of the stream which is a :class:`StreamMode` value.
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst]: I think we need to elaborate on this.  Add a sentence to list the enum values & explain what they mean.

File: Doc/library/asyncio-stream.rst [hunk 2]
```diff
@@ -367,13 +366,177 @@ Stream
...
+
+   .. method:: set_exception(exc)
+
+      Sets the exception to be raised when :meth:`read`, :meth:`readexactly` and
+      :meth:`readuntil` are called.
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst]: Ouch.  @asvetlov Why is this a public API?

File: Doc/library/asyncio-stream.rst [hunk 3]
```diff
@@ -367,13 +366,177 @@ Stream
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst:366]: Please update *high level API index* with the new APIs.

https://docs.python.org/3.8/library/asyncio-api-index.html

File: Doc/library/asyncio-stream.rst [hunk 4]
```diff
@@ -367,13 +366,177 @@ Stream
...
+
+   .. coroutinemethod:: sendfile(file, offset=0, count=None, *, fallback=True)
+
+      Calls :meth:`Stream.drain()` and rest of the arguments are passed directly to
+      :meth:`loop.sendfile`.
+
+   .. coroutinemethod:: start_tls(sslcontext, *, server_hostname=None, \
+                                  ssl_handshake_timeout=None)
+
+      Calls :meth:`Stream.drain()` and rest of the arguments are passed directly to
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst]: I'd actually elaborate a bit on this method - why we have it, ideally, with an example.

File: Lib/asyncio/streams.py [hunk 5]
```diff
@@ -70,6 +70,14 @@ def connect(host=None, port=None, *,
             server_hostname=None,
             ssl_handshake_timeout=None,
             happy_eyeballs_delay=None, interleave=None):
+    """Connect to TCP socket on *host* : *port* address and return a `Stream`
```
Comment (asvetlov) [inline Lib/asyncio/streams.py]: Minor: in docstrings for CPython we usually avoid markup. The docstring is for human reading, the documentation is generated from hand-written Sphinx pages.
`help()` doesn't render embedded markup too.
I know, older docstrings for the module contains back tickets at least.
Should we add new markup in this PR or avoid it in new updates?
@1st1 I need your opinion here

File: Lib/asyncio/streams.py [hunk 6]
```diff
@@ -108,6 +116,11 @@ def connect(host=None, port=None, *,
 
 
 def connect_read_pipe(pipe, *, limit=_DEFAULT_LIMIT):
+    """Takes a file-like object *pipe* to return a Stream object of the mode
```
Comment (1st1) [inline Lib/asyncio/streams.py]: Please make sure that all docstrings follow this: First paragraph is 80 characters long or less.  It should provide a quick summary of what the method/function is for.  Later paragraphs can elaborate on that further.

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Doc/library/asyncio-stream.rst (hunk 1 above)
Comment (asvetlov) [reply Doc/library/asyncio-stream.rst]: Agree, mentioning *read*, *write* and *read-write* modes make sense.
*Read* and *write* are used for pipes only, sockets are always read-write.

File: Doc/library/asyncio-stream.rst (hunk 2 above)
Comment (asvetlov) [reply Doc/library/asyncio-stream.rst]: Hard question.
It was public in the previous stream implementation.
We have two options:
1. Rename the method to `_set_exception()`.
2. Leave it as is but document that users should never touch it. Useless thing, I think (and a point of API misusage, like `Task.set_exception()`)

Comment (1st1) [reply Doc/library/asyncio-stream.rst]: I would like to rename the method to `_set_exception()` in 3.8.  If people come up with useful scenarios for that we can make it public in 3.9.  cc @ambv

File: Lib/asyncio/streams.py (hunk 5 above)
Comment (1st1) [reply Lib/asyncio/streams.py]: Using `*` for parameters improves readability IMO, using backticks isn't necessary (again, IMO).

File: Doc/library/asyncio-stream.rst (hunk 2 above)
Comment (asvetlov) [reply Doc/library/asyncio-stream.rst]: Preparing a PR.

Comment (asvetlov) [reply Doc/library/asyncio-stream.rst]: @tirkarthi please remove deprecated by #15762 methods and properties.
`asyncio.Stream` is a new class but it should support backward compatibility for the sake of asyncio subprocess subsystem.

There is no need to introduce deprecated-from-very-beginning API in docs.

File: Lib/asyncio/streams.py (hunk 5 above)
Comment (asvetlov) [reply Lib/asyncio/streams.py]: Ok, I have no strong opinion here.

Comment (tirkarthi) [reply Lib/asyncio/streams.py]: I have kept them and some of the other docstrings also have backticks to represent class names.

File: Doc/library/asyncio-stream.rst (hunk 1 above)
Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst]: Thanks, I added a bullet list of the values and a sentence about them.

File: Doc/library/asyncio-stream.rst (hunk 2 above)
Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst]: Thanks, removed `set_exception` and references to deprecated methods too in the docs as per the linked PR.

File: Doc/library/asyncio-stream.rst (hunk 3 above)
Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst:366]: Thanks, I have added the methods and classes added to the index. Can we also recommend in the deprecated methods about new replacements? Like `open_connection` could have a sentence that `connect` is recommended. I have used pretty much same sentences since they do the same and recommending would avoid confusion over duplicate functions.

File: Doc/library/asyncio-stream.rst (hunk 4 above)
Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst]: I have added a sentence over upgrading the TLS. I have also added similar sentence around `sendfile` above. I couldn't come up with a useful example than calling `connect` and then calling `start_tls` to upgrade SSL in next statement. If there is a more practical one please add in.

File: Lib/asyncio/streams.py (hunk 6 above)
Comment (tirkarthi) [reply Lib/asyncio/streams.py]: Done to here and other docstrings I have added to make sure first sentence have a short description.

File: Doc/library/asyncio-stream.rst (hunk 3 above)
Comment (1st1) [reply Doc/library/asyncio-stream.rst:366]: Yes! Please do that.

File: Doc/library/asyncio-api-index.rst [hunk 7]
```diff
@@ -132,18 +132,43 @@ High-level APIs to work with network IO.
...
+      -  Establish a TCP connection to send and receive data.
+
     * - ``await`` :func:`open_connection`
       -  Establish a TCP connection.
 
+    * - ``await`` :func:`connect_unix`
+      -  Establish a Unix socket connection to send and receive data.
+
     * - ``await`` :func:`open_unix_connection`
       -  Establish a Unix socket connection.
```
Comment (1st1) [inline Doc/library/asyncio-api-index.rst]: You can add "(deprecated in favor of the XXX function)" for old APIs.

File: Doc/library/asyncio-api-index.rst [hunk 8]
```diff
@@ -132,18 +132,43 @@ High-level APIs to work with network IO.
...
+      - Establish a connection to :term:`file-like object <file object>` *pipe*
+        to receive data.
+
+    * - :func:`connect_write_pipe`
+      - Establish a connection to :term:`file-like object <file object>` *pipe*
+        to send data.
+
+    * - :class:`Stream`
+      - High-level async/await object to send and receive network data.
+        It includes the API provided by :class:`StreamReader`
```
Comment (1st1) [inline Doc/library/asyncio-api-index.rst]: I'd rephrase to "Stream is a single object combining APIs of StreamReader and StreamWriter"

File: Doc/library/asyncio-stream.rst [hunk 9]
```diff
@@ -367,13 +366,167 @@ Stream
...
 
    Do not instantiate *Stream* objects directly; use API like :func:`connect`
    and :class:`StreamServer` instead.
 
    .. versionadded:: 3.8
 
+   .. attribute:: mode
+
+      Returns the mode of the stream which is a :class:`StreamMode` value. It could
+      be one of the below :
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst]:
```suggestion
      be one of the below:
```

File: Doc/library/asyncio-stream.rst [hunk 10]
```diff
@@ -367,13 +366,167 @@ Stream
...
+
+      * :attr:`StreamMode.READ` - Connection can receive data.
+      * :attr:`StreamMode.WRITE` - Connection can send data.
+      * :attr:`StreamMode.READWRITE` - Connection can send and receive data.
+
+   .. coroutinemethod:: abort()
+
+      Aborts the :ref:`Transport <asyncio-transport>` instance used by the Stream.
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst]: This is a relatively low-level description of what's happening here.  I'd use a more high-level description, such as "Aborts the connection immediately, without waiting for the send buffer to drain." (please double check that :))

File: Doc/library/asyncio-stream.rst [hunk 11]
```diff
@@ -367,13 +366,167 @@ Stream
...
+
+          stream.write(data)
+          await stream.drain()
+
+      This is a flow control method that interacts with the underlying
+      IO write buffer.  When the size of the buffer reaches
+      the high watermark, *drain()* blocks until the size of the
+      buffer is drained down to the low watermark and writing can
+      be resumed.  When there is nothing to wait for, the :meth:`drain`
+      returns immediately.
```
Comment (1st1) [inline Doc/library/asyncio-stream.rst:424]: This is a bit more subtle.  `drain()` is only needed when you do a few `stream.write()` calls.  When you have only one `stream.write()` you can just say `await stream.write()` and not bother with `drain()`.  @asvetlov am I right?

File: Doc/library/asyncio-api-index.rst (hunk 7 above)
Comment (tirkarthi) [reply Doc/library/asyncio-api-index.rst]: Thanks, done.

File: Doc/library/asyncio-api-index.rst (hunk 8 above)
Comment (tirkarthi) [reply Doc/library/asyncio-api-index.rst]: Thanks, reads better and combines both sentences. Reworded as suggested.

File: Doc/library/asyncio-stream.rst (hunk 10 above)
Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst]: I checked the method and it goes to call the abort method of the transport and waits for the close. I think it doesn't wait for data to be drained.

Comment (asvetlov) [reply Doc/library/asyncio-stream.rst]: Yes, `transport.abort()` clears pending write buffer if any,

File: Doc/library/asyncio-stream.rst (hunk 11 above)
Comment (asvetlov) [reply Doc/library/asyncio-stream.rst:424]: From my perspective, people always should use `await stream.write()` form and never `drain()` explicitly. There is no reason for late buffer overrun detection (and draining) if it can be done earlier.

The only reason why the method exists is backward compatibility for subprocess stdin/stdout/stderr pipes usage.

Comment (asvetlov) [reply Doc/library/asyncio-stream.rst:424]: I think we should reflect it in docs but sorry, I cannot propose exact wording.

File: Doc/library/asyncio-stream.rst [hunk 12]
```diff
@@ -367,13 +366,167 @@ Stream
...
+
+   .. coroutinemethod:: sendfile(file, offset=0, count=None, *, fallback=True)
+
+      Calls :meth:`Stream.drain()` to write to the connection and uses :meth:`loop.sendfile`
```
Comment (asvetlov) [inline Doc/library/asyncio-stream.rst]: `drain()` is implementation detail here.

```
Sends a *file* over the stream using an optimized syscall if available.

For other parameters meaning please see :meth:`AbstractEventloop.sendfile`.
```
or something like this.

File: Doc/library/asyncio-stream.rst [hunk 13]
```diff
@@ -367,13 +366,167 @@ Stream
...
+      to send *file* over *transport*.
+
+   .. coroutinemethod:: start_tls(sslcontext, *, server_hostname=None, \
+                                  ssl_handshake_timeout=None)
+
+      Calls :meth:`Stream.drain()` to write to the connection and uses :meth:`loop.start_tls`
```
Comment (asvetlov) [inline Doc/library/asyncio-stream.rst]: The same as for `sendfile`: the method upgrades a connection to TLS encrypted one.

File: Doc/library/asyncio-stream.rst [hunk 14]
```diff
@@ -367,13 +366,167 @@ Stream
...
+      is left in the internal buffer and can be read again.
+
+      If EOF is reached before the complete separator is found,
+      an :exc:`IncompleteReadError` exception is raised, and the internal
+      buffer is reset.  The :attr:`IncompleteReadError.partial` attribute
+      may contain a portion of the separator.
+
+   .. method:: write(data)
+
+      The method attempts to write the *data* to the underlying socket immediately.
```
Comment (asvetlov) [inline Doc/library/asyncio-stream.rst:514]: I think here and below for methods that can be used both with `await` and without it we should focus on `await` version first.  E.g. `await stream.write(data)` is more idiomatic than `stream.write(data); await stream.drain()`.

The first documentation line can be `Write *data* to the underlying socket; wait until the data is sent.`.

Thinking more, I believe we need to document two functions with the same name:
```
.. comethod:: write(data)

   Write *data* to the underlying socket; wait until the data is sent, e.g.::

      await stream.write(b'data')

.. method:: write(data)

   Long description here that teaches how combite `write()` with `drain()` and mentions that usually people should prefer `await stream,write()` version.
```
I'm curious is it supported by sphinx well or not? What version is a target for `:meth:Stream.write` expression?

Comment (tirkarthi) [reply Doc/library/asyncio-stream.rst:514]: It is supported. Here is a first pass version of this. The `Stream.write` target is set to the coroutine and the normal write method cannot be hyperlinked. We also do some explanation on the drain method so I have the docs as below : 

```rst
   .. coroutinemethod:: write(data)

      Write *data* to the underlying socket; wait until the data is sent, e.g.::

         await stream.write(data)

   .. method:: write(data)

      The method attempts to write the *data* to the underlying socket immediately.
      If that fails, the data is queued in an internal write buffer until it can be
      sent. :meth:`drain` can be used to flush the underlying buffer once writing is
      available::

         stream.write(data)
         await stream.drain()

      It is recommended to directly await on the `write()` method instead::

         await stream.write(data)
```

Screen Shot 2019-09-11 at 10 59 30 am

Comment (asvetlov) [reply Doc/library/asyncio-stream.rst:514]: Looks nice!

**Comment by tirkarthi** [comment]
I have pushed the changes to make sure `await stream.write(data)` as recommended with `stream.write(data); await stream.drain()` also documented with a method. I think the only detail pending now is over how multiple `stream.write` can be buffered to use `stream.drain` later as below that should be noted in the `stream.drain` docs.

https://github.com/python/cpython/pull/14488#discussion_r322740985

> This is a bit more subtle. drain() is only needed when you do a few stream.write() calls. When you have only one stream.write() you can just say await stream.write() and not bother with drain(). @asvetlov am I right?

**Comment by asvetlov** [comment]
> I think the only detail pending now is over how multiple stream.write can be buffered to use stream.drain later as below that should be noted in the stream.drain docs.

The interesting part is that multiple sync writes and single drain after doesn't help with buffering. Instead, it may consume more memory that expected.  

asyncio streams already have low and high watermarks for write buffer size control. If kernel write buffer is full the data is collected in asyncio internal write buffer. If the internal buffer size is greater than high watermark the writing is paused until the size of the buffer doesn't fall to less than the low watermark. After that write is resumed and keeps resumed until the high watermark is reached again.

So, data is effectively buffered even if `await stream.write()` notation is always used, `drain()` after muptiple sync `write()` calls doesn't make any value.

I doubt if we need document it here, in a method reference.  Maybe a separate asyncio section about buffering makes sense? 
The section is not related to streams itself, low-level `protocol.pause_writing()` / `protocol.resume_writing()` supports the same logic.

**Comment by 1st1** [comment]
Andrew, reading all of this makes me think about soft-deprecating non-await `write()` and `drain()` in the documentation (it will probably take us many releases before we can actually deprecate them, if ever).  This will actually encourage people to just write `await stream.write()` and not touch `drain()` etc.

**Comment by asvetlov** [comment]
> Andrew, reading all of this makes me think about soft-deprecating non-await write() and drain() in the documentation (it will probably take us many releases before we can actually deprecate them, if ever). This will actually encourage people to just write await stream.write() and not touch drain() etc.

@1st1 I told you at US PyCon that non-awaitable methods are useless and should be deprecated eventually. You had doubts, so we decided to support both.

I totally agree with the deprecation strategy.

Does soft-deprecation means `..deprecated:` section or completely removal these API from docs?
The second is more aggressive, sure. But it simplifies documentation writing and (much more important) reading a lot.

@1st1 what do you prefer?

**Comment by 1st1** [comment]
> Does soft-deprecation means ..deprecated: section or completely removal these API from docs?
The second is more aggressive, sure. But it simplifies documentation writing and (much more important) reading a lot.

I'd do it in a very soft way in 3.8/3.9; basically we just warn users that they are deprecated in the docs and that's it.

**Comment by asvetlov** [comment]
Ok, I agree with any decision.
@tirkarthi please add `.. deprecation::` sections.

Sorry for the endless list of notes. Writing a new stream API took me weeks, documenting it takes a long as well.

**Comment by tirkarthi** [comment]
Added deprecated notes to `Stream.drain` and method `Stream.write` to recommending usage of `await Stream.write` as a coroutine to be awaited. There is also `StreamWriter.write` and `StreamWriter.writelines` but they are already recommended to use `stream.write` and contain examples on using it in 3.7 where await is not supported so I have not added deprecations there.

No problem, I got to know better about the API while documentation and also closed some old and outdated issues related to Streams on the tracker with the PR. Appreciate your help on word formation and care to perfect user facing documentation.

**Review by asvetlov — CHANGES_REQUESTED** [review]
I like the current text, looks good.
The last nitpick: please provide an order to Stream methods.
I think alphabetical sorting is better than the current grouping.

**Comment by tirkarthi** [comment]
Good, done. The current order is alphabetic sort as below with attribute mode at first

```
mode

abort
at_eof
can_write_eof
close
drain
get_extra_info
is_closing
read
readexactly
readline
readuntil
sendfile
start_tls
wait_closed
write coroutine
write method
writelines
write_eof

``

**Review by asvetlov — APPROVED** [review]
Well done!
Thank you very much!

**Comment by asvetlov** [comment]
@1st1 please review again. The PR is ready for merging from my perspective

**Comment by tirkarthi** [comment]
Thank you very much Andrew and Yury. I can create a separate PR if there are additional improvements.

**Comment by asvetlov** [comment]
Oops. Sorry, @1st1 
I've added automerge label with the hope that the merge will be blocked until your approval.
@miss-islington has own logic for this situation.

Anyway, please feel free to express your objection if any, @tirkarthi and I can fix them in the following PR quickly.
